### PR TITLE
cmd: ex: standalone docker garbage collector

### PR DESCRIPTION
**Pull request by sjenning**
This implements a standalone docker GC daemon for use in OpenShift when using a non-docker runtime (e.g. cri-o) but using docker for builds.

https://trello.com/c/iO9d9R1w

```
$ openshift ex dockergc --help
Perform garbage collection to free space in docker storage 

If the OpenShift node is configured to use a container runtime other than docker, docker will still be used to do
builds.  However OpenShift itself will not manage the docker storage since it is not the container runtime for pods. 

This utility allows garbage collection to do be done on the docker storage.

Usage:
  openshift ex dockergc [NAME] [options]

Examples:
  # Perform garbage collection with the default settings
  openshift ex dockergc

Options:
      --dry-run=false: If true, show the result of the operation without performing it.
      --image-gc-high-threshold=80: The percent of disk usage after which image garbage collection is always run.
      --image-gc-low-threshold=60: The percent of disk usage before which image garbage collection is never run. Lowest
disk usage to garbage collect to.
      --minimum-ttl-duration=1h0m0s: Minimum age for a container or unused image before it is garbage collected.
Examples: '300ms', '10s' or '2h45m'.
  -o, --output='': Output results as yaml or json instead of executing, or use name for succint output (resource/name).
  -a, --show-all=false: When printing, show all resources (default hide terminated pods.)
      --show-labels=false: When printing, show all labels as the last column (default hide labels column)
      --sort-by='': If non-empty, sort list types using this field specification.  The field specification is expressed
as a JSONPath expression (e.g. '{.metadata.name}'). The field in the API resource specified by this JSONPath expression
must be an integer or a string.
      --template='': Template string or path to template file to use when -o=go-template, -o=go-template-file. The
template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].

Use "openshift options" for a list of global command-line options (applies to all commands).
```

@derekwaynecarr @vikaschoudhary16

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
can you add a yaml demonstrating how you would run this on cluster?

**Comment by sjenning** [comment]
@derekwaynecarr added the example daemonset spec and serviceaccount with required scc

**Comment by enj** [comment]
/unassign

Missing too much context to review 😄

**Comment by sjenning** [comment]
/retest

flake https://github.com/openshift/origin/issues/16414

File: examples/dockergc/dockergc-ds.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+          - "--image-gc-low-threshold=60"
+          - "--image-gc-high-threshold=80"
+          - "--minimum-ttl-duration=1h0m0s"
+          securityContext:
+            privileged: true
+          name: dockergc
+          resources:
+            requests:
+              memory: 30Mi
+              cpu: 100m
```
Comment (derekwaynecarr) [inline examples/dockergc/dockergc-ds.yaml]: can we size less?  maybe 50m?  how did you get to this number?

File: pkg/oc/experimental/dockergc/dockergc.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+	// collect to.
+	ImageGCLowThresholdPercent int32
+}
+
+var (
+	dockerGC_long = templates.LongDesc(`
+		Perform garbage collection to free space in docker storage
+
+		If the OpenShift node is configured to use a container runtime other than docker,
+		docker will still be used to do builds.  However OpenShift itself will not
```
Comment (derekwaynecarr) [inline pkg/oc/experimental/dockergc/dockergc.go]: s/will/may

File: pkg/oc/experimental/dockergc/dockergc.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+func (s oldestImagesFirst) Less(i, j int) bool { return s[i].Created < s[j].Created }
+
+// parseDockerTimestamp parses the timestamp returned by Interface from string to time.Time
+func parseDockerTimestamp(s string) (time.Time, error) {
+	// Timestamp returned by Docker is in time.RFC3339Nano format.
+	return time.Parse(time.RFC3339Nano, s)
+}
+
+func doGarbageCollection(ctx context.Context, client *dockerapi.Client, options *dockerGCConfigCmdOptions, rootDir string) error {
+	capacityBytes, usageBytes, err := getRootDirInfo(rootDir)
```
Comment (derekwaynecarr) [inline pkg/oc/experimental/dockergc/dockergc.go]: can we do a log at each gc interval?

Comment (derekwaynecarr) [reply pkg/oc/experimental/dockergc/dockergc.go]: it would be good to have something that logs before gathering root dir info

**Comment by derekwaynecarr** [comment]
@sjenning just a few nits.

**Comment by derekwaynecarr** [comment]
running out of disk space on a cri-o server would be a bug.

/kind bug

File: pkg/oc/experimental/dockergc/dockergc.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	dockerapi "github.com/docker/engine-api/client"
+	dockertypes "github.com/docker/engine-api/types"
+	dockerfilters "github.com/docker/engine-api/types/filters"
```
Comment (liggitt) [inline pkg/oc/experimental/dockergc/dockergc.go]: as long as these dependencies 1) already exist and 2) aren't ones we're trying to delete as part of the oc split (I don't think they are... @deads2k?), hanging this under experimental doesn't bother me

Comment (deads2k) [reply pkg/oc/experimental/dockergc/dockergc.go]: > as long as these dependencies 1) already exist and 2) aren't ones we're trying to delete as part of the oc split (I don't think they are... @deads2k?), hanging this under experimental doesn't bother me

We do want to see this leave as part of `oc`.  Is this command logically an `oc adm` command? If so, then `openshift ex` is ok and we'll move it to `oc ex` later.

File: examples/dockergc/dockergc-ds.yaml (hunk 1 above)
Comment (sjenning) [reply examples/dockergc/dockergc-ds.yaml]: copied it from node-exporter :smile:

File: pkg/oc/experimental/dockergc/dockergc.go (hunk 4 above)
Comment (derekwaynecarr) [reply pkg/oc/experimental/dockergc/dockergc.go]: @deads2k it should actually be `oc` but `openshift ex`.  the command is temporary and is avoiding us having to ship an official product image.

@sjenning i think /pkg/cmd/openshift experimental is the right home for this.  similar to those other experimental commands.

Comment (derekwaynecarr) [reply pkg/oc/experimental/dockergc/dockergc.go]: moved discussion to irc.  agree w/ @deads2k that we may want to handle this similar to kubernetes-namespace-reservation

Comment (derekwaynecarr) [reply pkg/oc/experimental/dockergc/dockergc.go]: @sjenning -- @deads2k and i spoke in person.  when `oc` moves out of tree, we will move this command with it.  at that point, we will need to make sure our ops team has the updated daemonset.

**Comment by sjenning** [comment]
flake #16870
/test extended_conformance_install_update

**Pull request merged**